### ISSUE TITLE
Ideas for pattern entry layout.

### ISSUE THREAD

**Issue by rkujawa** [grzegorz-kraszewski/loopmill#2]
I like the idea of having trakcer-style MIDI sequencer. You might want to take a look at how ReNoise did this (the demo version is fully functional in this area, there's Linux version too). It's a sample-based tracker, but has full MIDI support. The thing is that MIDI does not feel like an alien in it, it's almost as easy to use as samples.

http://renoise.com/

**Comment by grzegorz-kraszewski** [comment]
Loopmill is designed from the opposite point of view: it is MIDI sequencer first, then it can also include samples, probably by creating a virtual MIDI device using Paula. Solutions from ReNoise may be also hard to implement, because:
- ReNoise uses much larger screen resolution for the UI.
- ReNoise runs on much faster computers with lots of RAM. 

Considering the above, I'm open for detailed suggestions.

**Comment by rkujawa** [comment]
The basic ideas outlined in README.md seem pretty good for MIDI-first program. I'll add some more ideas in this issue (both ReNoise-inspired and my own):
- Pattern entry should support:
  - "Command" mode where copying, pasting, selecting the pattern possible, also calling some tracker functions via computer keyboard
  - "Edit" mode allowing changes to pattern (i.e. inserting notes, effects, etc.) 
  - "Record" mode allowing inserting data (notes particularly) into pattern in real time
- Allow entry into patter editor using:
  - MIDI keyboard in "record" mode - i.e. song timer is running, the pattern in moving, you can "improvise" on the keyboard
  - MIDI keyboard in "edit" mode - i.e. song stopped but pattern editing allowed
  - Computer keyboard in a classic tracker-like style during "record" mode
  - Computer keyboard in a classic tracker-like style during "edit" mode

**Comment by grzegorz-kraszewski** [comment]
I'm not sure if there is a need for separate "Command" and "Edit" states. I guess the idea behind it is to have all computer keyboard keys available in "Command" mode, while in "Edit" they emulate music keyboard. However ProTracker has no such two states.

Other than that, features listed are obvious to me.

**Comment by rkujawa** [comment]
Exactly, that would be the idea. When the amount of functions grow, it becomes difficult to access them all from keyboard, when most of keyboard is used to record/edit notes. When you are using both hands, one on the keyboard of the computer, the other on the MIDI keyboard, it's very annoying to have to reach for the mouse all the time to access tracker functions.

I know ProTrackes has no such two states, but implementing this 2-state mode shouldn't be difficult. Changing between Command/Edit mode could be done with a single press key. ReNoise does implement it that way.
